<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "https://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.9.1"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>libsigc++: exception_catch()</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="doxygen-extra.css" rel="stylesheet" type="text/css"/>
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">libsigc++
   &#160;<span id="projectnumber">3.2.0</span>
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.9.1 -->
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',false,false,'search.php','Search');
});
/* @license-end */</script>
<div id="main-nav"></div>
</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">exception_catch()<div class="ingroups"><a class="el" href="group__adaptors.html">Adaptors</a></div></div>  </div>
</div><!--header-->
<div class="contents">

<p><a class="el" href="namespacesigc.html#a087997a75b3111a3f3794d7215fd8cc1">sigc::exception_catch()</a> catches an exception thrown from within the wrapped functor and directs it to a catcher functor.  
</p>
<p><a class="el" href="namespacesigc.html#a087997a75b3111a3f3794d7215fd8cc1">sigc::exception_catch()</a> catches an exception thrown from within the wrapped functor and directs it to a catcher functor. </p>
<p>This catcher can then rethrow the exception and catch it with the proper type.</p>
<p>Note that the catcher is expected to return the same type as the wrapped functor so that normal flow can continue.</p>
<p>Catchers can be cascaded to catch multiple types, because uncaught rethrown exceptions proceed to the next catcher adaptor.</p>
<dl class="section user"><dt>Examples:</dt><dd><div class="fragment"><div class="line"><span class="keyword">struct </span>my_catch</div>
<div class="line">{</div>
<div class="line">  <span class="keywordtype">int</span> operator()()</div>
<div class="line">  {</div>
<div class="line">    <span class="keywordflow">try</span> { <span class="keywordflow">throw</span>; }</div>
<div class="line">    <span class="keywordflow">catch</span> (<a class="codeRef" href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a02969.html">std::range_error</a> e) <span class="comment">// catch what types we know</span></div>
<div class="line">      { <a class="codeRef" href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a01635.html#a7431d56d1e8cd7a9b854171294bd71c7">std::cerr</a> &lt;&lt; <span class="stringliteral">&quot;caught &quot;</span> &lt;&lt; e.<a class="codeRef" href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a02965.html#a45f5b9c4b7141ac429bf6ea99a7ff6bf">what</a>() &lt;&lt; <a class="codeRef" href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a01635.html#a8b3758bb8f17c440c7963363f42f69f3">std::endl</a>; }</div>
<div class="line">    <span class="keywordflow">return</span> 1;</div>
<div class="line">  }</div>
<div class="line">}</div>
<div class="line"><span class="keywordtype">int</span> foo(); <span class="comment">// throws std::range_error</span></div>
<div class="line"><a class="code" href="namespacesigc.html#a087997a75b3111a3f3794d7215fd8cc1">sigc::exception_catch</a>(&amp;foo, my_catch())();</div>
<div class="ttc" id="aa01635_html_a7431d56d1e8cd7a9b854171294bd71c7"><div class="ttname"><a href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a01635.html#a7431d56d1e8cd7a9b854171294bd71c7">std::cerr</a></div><div class="ttdeci">ostream cerr</div></div>
<div class="ttc" id="aa01635_html_a8b3758bb8f17c440c7963363f42f69f3"><div class="ttname"><a href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a01635.html#a8b3758bb8f17c440c7963363f42f69f3">std::endl</a></div><div class="ttdeci">basic_ostream&lt; _CharT, _Traits &gt; &amp; endl(basic_ostream&lt; _CharT, _Traits &gt; &amp;__os)</div></div>
<div class="ttc" id="aa02965_html_a45f5b9c4b7141ac429bf6ea99a7ff6bf"><div class="ttname"><a href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a02965.html#a45f5b9c4b7141ac429bf6ea99a7ff6bf">std::range_error::what</a></div><div class="ttdeci">virtual const char * what() const noexcept</div></div>
<div class="ttc" id="aa02969_html"><div class="ttname"><a href="http://gcc.gnu.org/onlinedocs/libstdc++/latest-doxygen/a02969.html">std::range_error</a></div></div>
<div class="ttc" id="anamespacesigc_html_a087997a75b3111a3f3794d7215fd8cc1"><div class="ttname"><a href="namespacesigc.html#a087997a75b3111a3f3794d7215fd8cc1">sigc::exception_catch</a></div><div class="ttdeci">decltype(auto) exception_catch(const T_functor &amp;func, const T_catcher &amp;catcher)</div><div class="ttdef"><b>Definition:</b> exception_catch.h:129</div></div>
</div><!-- fragment --></dd></dl>
<p>The functor <a class="el" href="namespacesigc.html#a087997a75b3111a3f3794d7215fd8cc1">sigc::exception_catch()</a> returns can be directly passed into <a class="el" href="classsigc_1_1signal__with__accumulator.html#a3765ae85a8ffaae0d48afa15fffe6009">sigc::signal::connect()</a>.</p>
<dl class="section user"><dt>Example:</dt><dd><div class="fragment"><div class="line">sigc::signal&lt;int()&gt; some_signal;</div>
<div class="line">some_signal.connect(<a class="code" href="namespacesigc.html#a087997a75b3111a3f3794d7215fd8cc1">sigc::exception_catch</a>(&amp;foo, my_catch));</div>
</div><!-- fragment --> </dd></dl>
</div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated on Fri Feb 11 2022 16:14:06 for libsigc++ by&#160;<a href="https://www.doxygen.org/index.html"><img class="footer" src="doxygen.svg" width="104" height="31" alt="doxygen"/></a> 1.9.1
</small></address>
</body>
</html>
